File: web/src/lib/mock/data/analytics.ts
import { registerRoute } from "../router";
import { STRATEGY_NAMES, ALL_SYMBOLS } from "./constants";
import { seededRandom } from "../generators";

// ---------------------------------------------------------------------------
// Strategy attribution data
// ---------------------------------------------------------------------------

interface StrategyAttribution {
  name: string;
  total_trades: number;
  win_rate: number;
  total_pnl: number;
  avg_win: number;
  avg_loss: number;
  profit_factor: number;
  max_consecutive_losses: number;
}

const strategyData: Record<string, StrategyAttribution> = {
  momentum_breakout: {
    name: "momentum_breakout",
    total_trades: 145,
    win_rate: 0.582,
    total_pnl: 12450,
    avg_win: 280,
    avg_loss: -195,
    profit_factor: 1.67,
    max_consecutive_losses: 5,
  },
  mean_reversion: {
    name: "mean_reversion",
    total_trades: 98,
    win_rate: 0.521,
    total_pnl: 4230,
    avg_win: 210,
    avg_loss: -180,
    profit_factor: 1.21,
    max_consecutive_losses: 7,
  },
  sentiment_alpha: {
    name: "sentiment_alpha",
    total_trades: 67,
    win_rate: 0.648,
    total_pnl: 8920,
    avg_win: 340,
    avg_loss: -220,
    profit_factor: 2.13,
    max_consecutive_losses: 3,
  },
  ml_ensemble: {
    name: "ml_ensemble",
    total_trades: 203,
    win_rate: 0.553,
    total_pnl: 15680,
    avg_win: 195,
    avg_loss: -155,
    profit_factor: 1.42,
    max_consecutive_losses: 6,
  },
  pairs_trading: {
    name: "pairs_trading",
    total_trades: 54,
    win_rate: 0.615,
    total_pnl: -2140,
    avg_win: 180,
    avg_loss: -320,
    profit_factor: 0.87,
    max_consecutive_losses: 8,
  },
  trend_following: {
    name: "trend_following",
    total_trades: 178,
    win_rate: 0.497,
    total_pnl: 3260,
    avg_win: 245,
    avg_loss: -200,
    profit_factor: 1.15,
    max_consecutive_losses: 9,
  },
};

const totalPnl = Object.values(strategyData).reduce(
  (sum, s) => sum + s.total_pnl,
  0,
);

// ---------------------------------------------------------------------------
// Correlation matrix
// ---------------------------------------------------------------------------

// First 7 symbols: AAPL, MSFT, GOOGL, AMZN, TSLA, NVDA, META
const correlationSymbols = ALL_SYMBOLS.slice(0, 7);

// Upper-triangle values (row-major, excluding diagonal)
// Order: AAPL(0), MSFT(1), GOOGL(2), AMZN(3), TSLA(4), NVDA(5), META(6)
const upperTriangle: Record<string, number> = {
  "0-1": 0.82, // AAPL-MSFT
  "0-2": 0.78, // AAPL-GOOGL
  "0-3": 0.62, // AAPL-AMZN
  "0-4": 0.42, // AAPL-TSLA
  "0-5": 0.75, // AAPL-NVDA
  "0-6": 0.72, // AAPL-META
  "1-2": 0.80, // MSFT-GOOGL
  "1-3": 0.65, // MSFT-AMZN
  "1-4": 0.38, // MSFT-TSLA
  "1-5": 0.79, // MSFT-NVDA
  "1-6": 0.74, // MSFT-META
  "2-3": 0.68, // GOOGL-AMZN
  "2-4": 0.45, // GOOGL-TSLA
  "2-5": 0.76, // GOOGL-NVDA
  "2-6": 0.77, // GOOGL-META
  "3-4": 0.48, // AMZN-TSLA
  "3-5": 0.58, // AMZN-NVDA
  "3-6": 0.55, // AMZN-META
  "4-5": 0.43, // TSLA-NVDA
  "4-6": 0.35, // TSLA-META
  "5-6": 0.85, // NVDA-META
};

function buildCorrelationMatrix(): number[][] {
  const n = correlationSymbols.length;
  const matrix: number[][] = Array.from({ length: n }, () =>
    Array(n).fill(0),
  );

  for (let i = 0; i < n; i++) {
    matrix[i][i] = 1.0;
    for (let j = i + 1; j < n; j++) {
      const val = upperTriangle[`${i}-${j}`] ?? 0.5;
      matrix[i][j] = val;
      matrix[j][i] = val;
    }
  }

  return matrix;
}

// ---------------------------------------------------------------------------
// Route handlers
// ---------------------------------------------------------------------------

// GET /api/analytics/attribution - strategy attribution report
registerRoute("GET", /^\/api\/analytics\/attribution$/, () => ({
  strategies: strategyData,
  total_pnl: totalPnl,
  best_strategy: "ml_ensemble",
  worst_strategy: "pairs_trading",
}));

// GET /api/analytics/monte-carlo - Monte Carlo simulation results
registerRoute("GET", /^\/api\/analytics\/monte-carlo$/, () => ({
  actual_final_value: 1042400,
  percentile: 67,
  median_simulated: 1028500,
  p5_simulated: 945200,
  p95_simulated: 1135800,
  worst_drawdown_p95: 0.128,
  n_simulations: 10000,
}));

// GET /api/analytics/drawdown - drawdown time-series
registerRoute("GET", /^\/api\/analytics\/drawdown$/, () => {
  const rng = seededRandom(321);
  const points: { index: number; drawdown_pct: number; value: number }[] = [];
  let value = 1000000;
  let peak = value;
  let drawdownPct = 0;

  for (let i = 0; i < 100; i++) {
    // Slowly increase value with noise
    const dailyReturn = 0.001 + (rng() - 0.52) * 0.025;
    value *= 1 + dailyReturn;

    // Track peak for drawdown calculation
    if (value > peak) {
      peak = value;
    }
    drawdownPct = ((peak - value) / peak) * 100;

    // Clamp to realistic range (0 to ~12%)
    drawdownPct = Math.min(drawdownPct, 12);

    points.push({
      index: i,
      drawdown_pct: Math.round(drawdownPct * 1000) / 1000,
      value: Math.round(value * 100) / 100,
    });
  }

  return { points };
});

// GET /api/analytics/correlation - symbol correlation matrix
registerRoute("GET", /^\/api\/analytics\/correlation$/, () => ({
  symbols: correlationSymbols,
  matrix: buildCorrelationMatrix(),
}));

File: src/dashboard/routers/analytics.py
"""Analytics endpoints: attribution, Monte Carlo, drawdown, correlation."""

from __future__ import annotations

from fastapi import APIRouter, Depends

from src.dashboard.dependencies import require_user, state
from src.db.models import UserRecord  # noqa: TC001

router = APIRouter(prefix="/api/analytics", tags=["analytics"])


def _build_attributed_fills():
    """Build attributed fills from portfolio fill history."""
    from src.analytics.models import AttributedFill

    if state.portfolio is None:
        return []

    fills = state.portfolio._fills
    attributed = []
    for fill in fills:
        # Try to determine strategy from DB records
        strategy = "unknown"
        attributed.append(AttributedFill(fill=fill, strategy=strategy))
    return attributed


@router.get("/attribution")
async def get_attribution(current_user: UserRecord = Depends(require_user)):  # noqa: B008
    """Full strategy attribution report."""
    from src.analytics.attribution import StrategyAttribution

    fills = _build_attributed_fills()

    # Try to get strategy info from DB
    if state.db and fills:
        trades = await state.db.list_trades(limit=1000)
        # Build a map of (symbol, side, approximate_time) -> strategy
        from src.analytics.models import AttributedFill

        strategy_map: dict[str, str] = {}
        for t in trades:
            strategy_map[t.id] = t.strategy

        # Re-attribute fills using trade records
        attributed = []
        trade_list = list(trades)
        for fill in state.portfolio._fills:
            strategy = "unknown"
            # Match by symbol and side
            for t in trade_list:
                if t.symbol == fill.symbol and t.side == fill.side.value:
                    strategy = t.strategy
                    trade_list.remove(t)
                    break
            attributed.append(AttributedFill(fill=fill, strategy=strategy))
        fills = attributed

    analyzer = StrategyAttribution()
    report = analyzer.analyze(fills)

    return {
        "strategies": {
            name: {
                "name": stats.name,
                "total_trades": stats.total_trades,
                "win_rate": stats.win_rate,
                "total_pnl": stats.total_pnl,
                "avg_win": stats.avg_win,
                "avg_loss": stats.avg_loss,
                "profit_factor": stats.profit_factor,
                "max_consecutive_losses": stats.max_consecutive_losses,
            }
            for name, stats in report.strategies.items()
        },
        "total_pnl": report.total_pnl,
        "best_strategy": report.best_strategy,
        "worst_strategy": report.worst_strategy,
    }


@router.get("/monte-carlo")
async def get_monte_carlo(current_user: UserRecord = Depends(require_user)):  # noqa: B008
    """Monte Carlo simulation results with percentile curves."""
    from src.analytics.attribution import StrategyAttribution
    from src.analytics.monte_carlo import MonteCarloSimulator

    fills = _build_attributed_fills()
    if not fills:
        return {
            "actual_final_value": 100000,
            "percentile": 50,
            "median_simulated": 100000,
            "p5_simulated": 100000,
            "p95_simulated": 100000,
            "worst_drawdown_p95": 0,
            "n_simulations": 0,
            "equity_curves": {},
        }

    analyzer = StrategyAttribution()
    analyzer.analyze(fills)

    # Gather all trades from all strategies
    all_trades = []
    for strategy_fills in _group_fills_by_strategy(fills).values():
        from src.analytics.attribution import _pair_fills

        all_trades.extend(_pair_fills(strategy_fills))

    initial_cash = 100000.0
    if state.portfolio:
        initial_cash = float(state.portfolio._cash) + sum(
            float(f.fill_price * f.quantity)
            if f.side.value == "buy"
            else -float(f.fill_price * f.quantity)
            for f in state.portfolio._fills
        )

    simulator = MonteCarloSimulator(n_simulations=500, seed=42)
    result = simulator.simulate(all_trades, initial_cash)

    return {
        "actual_final_value": result.actual_final_value,
        "percentile": result.percentile,
        "median_simulated": result.median_simulated,
        "p5_simulated": result.p5_simulated,
        "p95_simulated": result.p95_simulated,
        "worst_drawdown_p95": result.worst_drawdown_p95,
        "n_simulations": result.n_simulations,
    }


@router.get("/drawdown")
async def get_drawdown(current_user: UserRecord = Depends(require_user)):  # noqa: B008
    """Drawdown curve data."""
    if state.portfolio is None:
        return {"points": []}

    fills = state.portfolio._fills
    if not fills:
        return {"points": []}

    # Build equity curve
    initial = 100000.0
    equity = [initial]
    running = initial
    for fill in fills:
        impact = float(fill.fill_price * fill.quantity)
        if fill.side.value == "sell":
            running += impact
        else:
            running -= impact
        equity.append(running)

    # Compute drawdown curve
    peak = equity[0]
    drawdown_points = []
    for i, value in enumerate(equity):
        if value > peak:
            peak = value
        dd = (peak - value) / peak * 100 if peak > 0 else 0
        drawdown_points.append(
            {
                "index": i,
                "value": value,
                "drawdown_pct": dd,
            }
        )

    return {"points": drawdown_points}


@router.get("/correlation")
async def get_correlation(current_user: UserRecord = Depends(require_user)):  # noqa: B008
    """Correlation matrix between symbols."""
    if state.db is None:
        return {"symbols": [], "matrix": []}

    # Get OHLC data for correlation calculation
    bars = await state.db.query_ohlc_bars(limit=5000)
    if not bars:
        return {"symbols": [], "matrix": []}

    # Group closes by symbol
    from collections import defaultdict

    closes: dict[str, list[float]] = defaultdict(list)
    for bar in bars:
        closes[bar.symbol].append(float(bar.close))

    symbols = sorted(closes.keys())
    if len(symbols) < 2:
        return {"symbols": symbols, "matrix": [[1.0]]}

    # Simple correlation calculation
    import math

    def correlation(x: list[float], y: list[float]) -> float:
        n = min(len(x), len(y))
        if n < 2:
            return 0.0
        x, y = x[:n], y[:n]
        mx = sum(x) / n
        my = sum(y) / n
        sx = math.sqrt(sum((xi - mx) ** 2 for xi in x) / n)
        sy = math.sqrt(sum((yi - my) ** 2 for yi in y) / n)
        if sx == 0 or sy == 0:
            return 0.0
        cov = sum((xi - mx) * (yi - my) for xi, yi in zip(x, y, strict=False)) / n
        return cov / (sx * sy)

    matrix = []
    for s1 in symbols:
        row = []
        for s2 in symbols:
            if s1 == s2:
                row.append(1.0)
            else:
                row.append(round(correlation(closes[s1], closes[s2]), 4))
        matrix.append(row)

    return {"symbols": symbols, "matrix": matrix}


def _group_fills_by_strategy(fills):
    from collections import defaultdict

    grouped = defaultdict(list)
    for f in fills:
        grouped[f.strategy].append(f)
    return grouped
